pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -18,158 +18,219 @@ package credentials
 	"k8s.io/kubernetes/pkg/version"
 )
 
-const awsChinaRegionPrefix = "cn-"
-const awsStandardDNSSuffix = "amazonaws.com"
-const awsChinaDNSSuffix = "amazonaws.com.cn"
-const registryURLTemplate = "*.dkr.ecr.%s.%s"
-
-// awsHandlerLogger is a handler that logs all AWS SDK requests
-// Copied from pkg/cloudprovider/providers/aws/log_handler.go
-func awsHandlerLogger(req *request.Request) {
-	service := req.ClientInfo.ServiceName
-	region := req.Config.Region
-
-	name := "?"
-	if req.Operation != nil {
-		name = req.Operation.Name
-	}
+// init registers a credential provider for each registryURLTemplate and creates
+// an ECR token getter factory with a new cache to store token getters
+func init() {
+	credentialprovider.RegisterCredentialProvider("amazon-ecr",
+		newECRProvider(&ecrTokenGetterFactory{cache: make(map[string]tokenGetter)}))
+}
 
-	klog.V(3).Infof("AWS request: %s:%s in %s", service, name, *region)
+// ecrProvider is a DockerConfigProvider that gets and refreshes tokens
+// from AWS to access ECR.
+type ecrProvider struct {
+	cache         cache.Store
+	getterFactory tokenGetterFactory
 }
 
-// An interface for testing purposes.
-type tokenGetter interface {
-	GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error)
+var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
+
+func newECRProvider(getterFactory tokenGetterFactory) *ecrProvider {
+	return &ecrProvider{
+		cache:         cache.NewExpirationStore(stringKeyFunc, &ecrExpirationPolicy{}),
+		getterFactory: getterFactory,
+	}
 }
 
-// The canonical implementation
-type ecrTokenGetter struct {
-	svc *ecr.ECR
+// Enabled implements DockerConfigProvider.Enabled. Enabled is true if AWS
+// credentials are found.
+func (p *ecrProvider) Enabled() bool {
+	sess, err := session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		klog.Errorf("while validating AWS credentials %v", err)
+		return false
+	}
+	if _, err := sess.Config.Credentials.Get(); err != nil {
+		klog.Errorf("while getting AWS credentials %v", err)
+		return false
+	}
+	return true
 }
 
-func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error) {
-	return p.svc.GetAuthorizationToken(input)
+// LazyProvide is lazy
+// TODO: PR to remove LazyProvide
+func (p *ecrProvider) LazyProvide(repoToPull string) *credentialprovider.DockerConfigEntry {
+	return nil
 }
 
-// lazyEcrProvider is a DockerConfigProvider that creates on demand an
-// ecrProvider for a given region and then proxies requests to it.
-type lazyEcrProvider struct {
-	region         string
-	regionURL      string
-	actualProvider *credentialprovider.CachingDockerConfigProvider
+// Provide returns a DockerConfig with credentials from the cache if they are
+// found, or from ECR
+func (p *ecrProvider) Provide(repoToPull string) credentialprovider.DockerConfig {
```

Comment:
I think this method can read better like this:
```go
func (p *ecrProvider) Provide(image string) credentialprovider.DockerConfig {
        parsed, err := parseRepoUrl(image)
        if err != nil {
               klog.Errorf("while parsing ECR repository URL %v", err)
               return credentialprovider.DockerConfig{}
        }    
    
        cfg, exists := gp.getFromCache(parsed)
        if exists {
                 klog.V(6).Infof("Got ECR credentials from cache for %s", parsed.registry)
                 return cfg
        }

         klog.V(3).Info("unable to get ECR credentials from cache, checking ECR API")

         cfg, err := p.getFromECR(parsed)
         if err != nil {
                klog.Errorf("unable to get credentials from ECR for %s %v", parsed.registry, err)
                return credentialprovider.DockerConfig{}
         }

         klog.V(3).Infof("Got ECR credentials from ECR API for %s", parsed.registry)
         return cfg
}
```

Reply:
The cache hit code path should be higher verbosity IMO

Reply:
Making it a tiny bit different, lmk if it is ok.

Reply:
lgtm!